kafka-rest/src/main/java/io/confluent/kafkarest/resources/v3/ProduceAction.java

```diff
@@ -92,9 +128,46 @@ public void produce(
       @PathParam("topicName") String topicName,
       MappingIterator<ProduceRequest> requests)
       throws Exception {
-    ProduceController controller = produceController.get();
-    StreamingResponse.from(requests)
-        .compose(request -> produce(clusterId, topicName, request, controller))
+
+    if (rateLimit()) {
```

Comment:
For your first point:
If the rest-producer rate limit ends up being lower than the overall limit then we want this check in place too don't we?  We shouldn't just rely on the overall limit for produces in case we need to set the limit really low.  Especially as it is possible to call produce in a non-streaming way.

For your second point:
The intent of having this block of code was to fail fast, before we even make a produceController, in the case where the produce endpoint is getting hit with lots of individual requests. It doesn't matter if there are multiple messages in the requests MappingIterator (ie in the case where it is a streaming produce, rather than individual calls) - that flow is picked up by the second block of rate limiting code.   The intent of this code is to block too many individual produce calls. That means I don't think we want this check hanging off the .from(requests) part here, it should be at the top level.